### ISSUE TITLE
Retaining Backwards Compatibility when Splitting Existing Package into Library and Executable Package

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#331]
Assume that the package `xxx` exports the qualified targets `exe{xxx}` and `lib{xxx}`. According to the manuals and the HOWTOs, there are multiple reasons to split the `xxx` package into the separate packages `libxxx` and `xxx` that export `lib{xxx}` and `exe{xxx}`, respectively. For example, as the library and the executable get more complex they start to require different dependencies.

Up to that point, each consumer had to use `depends: xxx ^1.0.0` in the `manifest` and `import xxx = xxx%lib{xxx}` in a `buildfile` of their package to correctly import the `xxx` library. By introducing the above described package split, we would break the code of all consumers as they would now need to use `depends: libxxx ^1.0.0` and `import xxx = libxxx%lib{xxx}`. Even though this would be a minor change on the consumer's side, it is still a breaking change and every consumer would need to do that.

To retain backwards compatibility, I suggest to re-export the library `lib{xxx}` by using a binless wrapper library in the `xxx` package after importation from the `libxxx` package. The idea is shown by the following `buildfile` snippet under the assumption of a default `export.build` file and has been successfully tested with [this repository](https://github.com/lyrahgames/build2-lib-reexport-test).

```
import xxx = libxxx%lib{xxx}

./: lib{xxx}: $xxx
{
  cxx.export.libs = $xxx
  bin.lib.prefix = # No prefix
}

./: exe{xxx}: {hxx ixx txx cxx}{**} $xxx testscript
```

I was mainly concerned with the installation of `pkg-config` files because the re-exported library is called the same way. To circumvent any overwrites, I removed the library prefix for the binless re-export wrapper library. In such a way, its `pkg-config` files are called `xxx[.static|.shared].pc` and not `libxxx[.static|.shared].pc`. I am quite sure that there is a better and more direct way to achieve the `pkg-config` file renaming. According to the [build2 Build System Manual: Importation of Installed Libraries](https://build2.org/stage/build2/doc/build2-build-system-manual.xhtml#cc-import-installed), for each consumer, that still uses `depends: xxx ^1.0.0` and `import xxx = xxx%lib{xxx}`, build2 will directly import the actual `lib{xxx}` library in the installed case. As of that, maybe disabling `pkg-config` file generation for the binless wrapper or its installation at all would be even better.

What do you think about this solution? Are there better ways to do that? Did I forget anything?

Furthermore, with this approach, it would also be easy to add the `lib-` prefix to library packages that don't use it. For example, `Eigen` and `doctest` packages could be renamed to `libEigen` and `libdoctest`. However, we could still use the packages `Eigen` and `doctest` to re-export the libraries for backwards compatibility. In such a way, we would gain a little more consistency. What do you think about that?

It would be nice to document a good approach in the [build2 Packaging Guide](https://build2.org/stage/build2-toolchain/doc/build2-toolchain-packaging.xhtml). If you wish, I could write something.

**Comment by boris-kolpackov** [comment]
I believe the correct time to split the library out of a combined project is when you are about to have the first external consumer. In other words, you keep the library combined with the executable until you want to use it somewhere else. Failed that, I think the next best thing is to call it `libxxx2` so that old consumers continue using the old import. I think the metadata library is a clever idea but it is complex and is bound to have corner cases, like that `pkg-config` file clash (I wouldn't be surprised if there were others).

> I am quite sure that there is a better and more direct way to achieve the `pkg-config` file renaming.

No, you are actually not supposed to renames them. You can read me on what happens when there is no consistent naming here: https://build2.org/stage/build2/doc/build2-build-system-manual.xhtml#cc-import-installed

> Furthermore, with this approach, it would also be easy to add the lib- prefix to library packages that don't use it. For example, Eigen and doctest packages could be renamed to libEigen and libdoctest. However, we could still use the packages Eigen and doctest to re-export the libraries for backwards compatibility. In such a way, we would gain a little more consistency. What do you think about that?

Yes, it's tempting but I think the extra complexity and variability (and the associated confusion for the user) is not worth it.

> It would be nice to document a good approach in the [build2 Packaging Guide](https://build2.org/stage/build2-toolchain/doc/build2-toolchain-packaging.xhtml).

Yes, though it seems the suggested approach so far is "don't" ;-)

**Comment by boris-kolpackov** [comment]
> Specifically, refer to [the build2 package for Lua](https://github.com/build2-packaging/lua) and https://github.com/build2-packaging/lua/issues/2.
> I would like to address this issue and did not want to break backwards compatibility.

Yes, mistakes like these do throw a wrench into my simple plan, unfortunately.

Perhaps for cases like these we could have a simpler workaround:

1. We could allow the export stub (`export.build`) to return a project qualified target which we will then try to import.

2. When splitting `lua` into two packages, you add the extra code to `lua`'s export stub to translate `lua%lib{lua}` to `liblua%lib{lua}`.

While I cannot guarantee this approach has no corner cases, I think it had a better chance of working out. 

In fact, one thing that is of concern is that we will now be resolving an import of `liblua%` from a project that depends on `lua` and will therefore may not have the necessarily `config.import.liblue` value (`bpkg` will only pass such value for dependencies). 

But maybe there is an even simpler way: instead of returning `liblua%lib{lua}` from `lua`'s `export.build`, we can import it and return the actual target. Specifically, the relevant files in the new `lua` package could look like this:

```
# buildfile
import! [rule_hint=c.link] liblua = liblua%lib{lua}

...
```

```
# export.build

$out_root/
{
  include ./
}

if ($import.target == lib{lua}  || \
    $import.target == liba{lua} || \
    $import.target == libs{lua})
{
  export $directory($($out_root/ liblua))/$import.target
}
else
{
  export $out_root/$import.target
}
```

I haven't tested this but I think it should do the trick.
